sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala

```diff
@@ -275,6 +279,7 @@ private[hive] class SparkExecuteStatementOperation(
   override def cancel(): Unit = {
     logInfo(s"Cancel '$statement' with $statementId")
     cleanup(OperationState.CANCELED)
+    HiveThriftServer2.listener.onStatementCanceled(statementId)
```

Comment:
This can call `onStatementCanceled()` before the thread launching `execute()` gets to `onStatementStarted()`, which will result in the listener throwing an NoSuchElementException on executionList(id).
`onStatementStarted()` could be moved to `runInternal()` in the main thread to prevent that, but then the `catch` block there has to add `onStatementError()`. It then would also be good to extend the `try` block to the start of `execute()`

Reply:
This can truly happen, especially when running SQL asynchronously， I add a new STATUS of ExecutStatus.PREPARED  in runInternal(), then onStatementStart just change  status, this can fix this problem.

Reply:
add a new status ExecutionState.PREPARED to fix this problem, call onStatementPrepared in runInternal().